File: Polynomial Optimization/heteroclinic_LowerBnd.m
% -------------------------------------------------------------------------
% Lower Bounds on the Existence of a Heteroclinic Orbit
%
% The goal of this script is to use semidefinite programming to provide
% lower bounds on the range of mu values for which a heteroclinic 
% connection exists in the system
%      x1' = x2, 
%      x2' = -mu*x2 - x1*(1-x1).
% The script searches for a desired auxiliary function using SOS 
% relaxations and localizations in phase space.   
%
% To run this notebook one requires paths to the freely available software
% packages YALMIP and MOSEK. YALMIP can be downloaded at:
%           https://yalmip.github.io/download/
% and MOSEK can be downloaded at:
%           https://www.mosek.com/downloads/
%
% This script accompanies Section 4.2 of Data-Driven Methods for
% Dynamic Systems. 
%
% Author: Jason J. Bramburger
% -------------------------------------------------------------------------

% Clean workspace
clear all; 
close all;
yalmip clear
clc

%% Bisection Method for Finding Upper Bound

format long

% ODE mu parameter
m = 2; 

% Bounding method parameters
degV = 20; % Degree of auxiliary function
lambda = 10^3;

%Bisection Method
muleft = 0;
muright = 2;
while abs(muright - muleft) >= 1e-5

    mumid = 0.5*(muleft + muright);
    flag = volume(mumid,m,degV,lambda);
    
    if flag == 0
       muleft = mumid; 
    else
       muright = mumid;
    end
    
    mumid
end

%Print Results
fprintf('The lower bound on the minimum speed for m = %d is %f found using degree %d polynomials.\n',m,mumid,degV)

%%
function flag = volume(mu,m,d,lambda)

% SDP variables
sdpvar u v

% Epsilon value
eps = 1e-4;

% Auxiliary function
[V, cV] = polynomial([u v], d);

% S Procedure
d2 = d+m;
[s1, c1] = polynomial([u v], d2);
[s2, c2] = polynomial([u v], d2);
[s3, c3] = polynomial(u, d2);

% Derivatives
dVdu = jacobian(V,u);
dVdv = jacobian(V,v);

% Replacements
Vv0 = replace(V,v,0);

%Constraints
cons = [];
cons = [cons, replace(V,[u v], [0 -eps]) == 0, replace(V, [u v], [1 0]) == 0];
cons = [cons, sos(lambda*(dVdu*v - mu*dVdv*v - dVdv*(1-u)*u^m) - V - u*(1-u)*s1 + v*s2)]; 
cons = [cons, sos(-Vv0 - eps*(1-u) - u*(1-u)*s3)];
cons = [cons, sos(s1), sos(s2), sos(s3)];

%SOS Solver
ops = sdpsettings('solver','mosek','verbose',0,'cachesolvers',1);
sol = solvesos(cons,[],ops,[cV;c1; c2; c3]);

%Return whether solvesos failed or succeeded
flag = sol.problem;

end

File: Data-Driven Polynomial Optimization/Julia Scripts/heteroclinic_LowerBnd.jl
using DynamicPolynomials, SumOfSquares, MosekTools
include("utils.jl")

# ODE mu parameter
m = 2

# Bounding method parameters
degV = 20  # Degree of auxiliary function
λ = 10^3

# Bisection Method for Finding Upper Bound
function volume(μ, m, d, λ)
    @polyvar u v

    ϵ = 1e-4;

    model = SOSModel(Mosek.Optimizer)
    set_silent(model)

    #Auxiliary function
    @variable(model, V, createPoly([u,v], d))

    # S procedure
    d2 = d + m 
    @variables(model, begin
        s1, createPoly([u,v], d2)
        s2 ,createPoly([u,v], d2)
        s3 , createPoly(u, d2)
    end)

    # Derivatives
    dVdu = differentiate(V, u)
    dVdv = differentiate(V, v)

    # Replacements
    Vv0 = subs(V, v=>0)

    @constraints(model, begin
        subs(V,u=>0, v=>-ϵ) == 0
        subs(V,u=>1,v=>0) == 0
        λ*(dVdu*v - μ*dVdv*v - dVdv*(1-u)*u^m) -V -u*(1-u)*s1 + v*s2 in SOSCone()
        -Vv0 -ϵ*(1-u) -u*(1-u)*s3 in SOSCone()
        s1 in SOSCone()
        s2 in SOSCone()
        s3 in SOSCone()
    end)

    optimize!(model)

    # Return whether we are able to solve the problem or not
    return primal_status(model)
end

#Bisection method
μ_left = 0
μ_right = 2
μ_mid = 0

while abs((μ_right - μ_left)) >= 1e-5
    μ_mid = 0.5 * (μ_left + μ_right)
    flag = volume(μ_mid, m, degV, λ)

    if flag==FEASIBLE_POINT
        μ_left = μ_mid
    else
        μ_right = μ_mid
    end
end

println("The lower bound on the minimum speed for m = $m is $μ_mid found using degree $degV polynomials.\n")